contracts/LockedNORI.sol

```diff
@@ -40,127 +63,339 @@ contract LockedNORI is ERC777Upgradeable, ERC20PresetMinterPauserUpgradeable, ER
     __Pausable_init_unchained();
     __ERC20Pausable_init_unchained();
     __ERC20PresetMinterPauser_init_unchained("Locked NORI", "lNORI");
-    __ERC20Wrapper_init_unchained(IERC20Upgradeable(address(noriAddress)));
-    __ERC777_init_unchained("Locked NORI", "lNORI", new address[](0));
+    address[] memory operators = new address[](1);
+    operators[0] = _msgSender();
+    __ERC777_init_unchained("Locked NORI", "lNORI", operators);
     _setupRole(DEFAULT_ADMIN_ROLE, _msgSender());
+    _underlying = ERC777Upgradeable(address(noriAddress));
+    _erc1820 = IERC1820RegistryUpgradeable(0x1820a4B7618BdE71Dce8cdc73aAB6C95905faD24); // todo
+    _erc1820.setInterfaceImplementer(address(this), keccak256("ERC777TokensRecipient"), address(this));
+  }
+
+  function tokensReceived(
+    address,
+    address,
+    address,
+    uint256 amount,
+    bytes calldata userData,
+    bytes calldata operatorData
+  ) external override {
+    require(
+      msg.sender == address(_underlying),
+      "LockedNORI: This contract can only receive NORI"
+    ); // todo verify this can only be invoked by the nori contract
+    // todo restrict such that only admin can invoke this function
+    _depositFor(
+      amount,
+      userData,
+      operatorData
+    );
   }
 
   /**
-   * @dev grantTo: Sets up an unlock schedule for recipient.
-   * Tokens must be wrapped in the name of the recipient separately using `depositFor`.
-   * This allows us to start the unlock schedule running for a recipient who may
-   * not yet have vested into all of their granted tokens.
+   * @dev Internal callback from tokensDeposited hook, wraps minting wrapped token and grant setup.
    */
-  function grantTo(
-    address recipient,
+  function _depositFor(
     uint256 amount,
-    uint256 startTime,
-    uint256 vestEndTime,
-    uint256 unlockEndTime,
-    uint256 cliff1Time,
-    uint256 cliff2Time,
-    uint256 vestCliff1Amount,
-    uint256 vestCliff2Amount,
-    uint256 unlockCliff1Amount,
-    uint256 unlockCliff2Amount
-  ) external onlyRole(TOKEN_GRANTER_ROLE) whenNotPaused {
+    bytes calldata userData,
+    bytes calldata operatorData
+  ) internal returns (bool) {
+    (address recipient, uint256 startTime) = abi.decode(userData, (address, uint256));
+    // TODO:SW Make this more explicit
+    // If a startTime parameter is non-zero then set up a schedule
+    if (startTime > 0) {
+      _grantTo(
+        amount,
+        userData
+      );
+    }
+
+    ERC777Upgradeable._mint(
+      recipient,
+      amount,
+      userData,
+      operatorData
+    );
+    return true;
+  }
+
+  /**
+   * @dev Allow a user to burn a number of wrapped tokens and withdraw the corresponding number of _underlying tokens.
+   */
+  function withdrawTo(address account, uint256 amount)
+    public
+    virtual
+    returns (bool)
+  {
+    TokenGrant storage grant = grants[account];
```

Comment:
does this need to check that they can claim the amount specified or is that happening somehow elsewhere?

e.g.,

```solidity
require(grant.originalAmount - grant.claimedAmount >= amount, "LockedNORI: insufficient remaining token amount");
```

Reply:
also, should it be updating the `grant.grantAmount` ?

Reply:
I looked into inverting this logic but it makes it more confusing to rely on `remainingBalance` given how the schedule library is designed.